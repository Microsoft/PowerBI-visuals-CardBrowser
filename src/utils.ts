/*
 * Copyright 2018 Uncharted Software Inc.
 */

import DataView = powerbi.DataView;
import DataViewMetadataColumn = powerbi.DataViewMetadataColumn;

import * as $ from 'jquery';

/**
 * Finds and returns the dataview column(s) that matches the given data role name.
 *
 * @param  {DataView} dataView     A Powerbi dataView object.
 * @param  {string}   dataRoleName A name of the role for the columen.
 * @param  {boolean}  multi        A boolean flag indicating whether to find multiple matching columns.
 * @return {any}                   A dataview table column or an array of the columns.
 */
export function findColumn(dataView: DataView | any, dataRoleName: string, multi?: boolean): any {
    const columns = dataView.metadata.columns;
    const result = (columns || []).filter((col: any) => col && col.roles[dataRoleName]);
    return multi
        ? (result[0] && result)
        : result[0];
}

/**
 * Check if provided dataView has all the columns with given data role names.
 *
 * @export
 * @param   {DataView} dataView      A Powerbi dataView object.
 * @param   {string[]} dataRoleNames An array of the data role names for corresponding columns.
 * @returns {boolean}                A Boolean value indicating whether the dataView has all matching columns.
 */
export function hasColumns(dataView: DataView | any, dataRoleNames: string[]): boolean {
    return dataRoleNames.reduce((prev, dataRoleName) => prev && findColumn(dataView, dataRoleName) !== undefined, true);
}

/**
 * Check if given column has a given role.
 */
export function hasRole(column: DataViewMetadataColumn, roleName: string) {
    return Boolean(column  && column.roles[roleName]);
}

// https://stackoverflow.com/questions/35962586/javascript-remove-inline-event-handlers-attributes-of-a-node#35962814
export function removeScriptAttributes(el) {
    const attributes = [].slice.call(el.attributes);

    for (let i = 0; i < attributes.length; i++) {
        const att = attributes[i].name;

        if (att.indexOf('on') === 0) {
            el.attributes.removeNamedItem(att);
        }
    }
}

/**
 * Removes dangerous tags, such as scripts, from the given HTML content.
 * @param {String} html - HTML content to clean
 * @param {Array} whiteList - Array of HTML tag names to accept
 * @returns {String} HTML content, devoid of any tags not in the whitelist
 */
export function sanitizeHTML(html: string, whiteList: string[]): string {
    let cleanHTML = '';
    if (html && whiteList && whiteList.length) {
        // Stack Overflow is all like NEVER PARSE HTML WITH REGEX
        // http://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags/1732454#1732454
        // plus the C# whitelist regex I found didn't work in JS
        // http://stackoverflow.com/questions/307013/how-do-i-filter-all-html-tags-except-a-certain-whitelist#315851
        // So going with the innerHTML approach...
        // http://stackoverflow.com/questions/6659351/removing-all-script-tags-from-html-with-js-regular-expression

        let doomedNodeList = [];

        if (!document.createTreeWalker) {
            return ''; // in case someone's hax0ring us?
        }

        let div = $('<div/>');

        // For the reasons above, we do need innerHTML, so suppress tslint
        // tslint:disable-next-line
        div.html(html);

        let filter: any = function (node) {
            removeScriptAttributes(node);
            if (whiteList.indexOf(node.nodeName.toUpperCase()) === -1) {
                return NodeFilter.FILTER_ACCEPT;
            }

            return NodeFilter.FILTER_SKIP;
        };

        filter.acceptNode = filter;

        // Create a tree walker (hierarchical iterator) that only exposes non-whitelisted nodes, which we'll delete.
        let treeWalker = document.createTreeWalker(
            div.get()[0],
            NodeFilter.SHOW_ELEMENT,
            filter,
            false
        );

        while (treeWalker.nextNode()) {
            doomedNodeList.push(treeWalker.currentNode);
        }

        let length = doomedNodeList.length;
        for (let i = 0; i < length; i++) {
            if (doomedNodeList[i].parentNode) {
                try {
                    doomedNodeList[i].parentNode.removeChild(doomedNodeList[i]);
                } catch (ex) { }
            }
        }

        // convert back to a string.
        cleanHTML = div.html().trim();
    }

    return cleanHTML;
}


/**
 * Converts from RGB color space to HSL color space.
 *
 * @method toHSL
 * @param {Array} rgb - An array containing the RGB components.
 * @returns {Array}
 */
function toHSL(rgb) {
    const [r, g, b] = rgb.map(n => n / 255);

    const max = Math.max(r, g, b);
    const min = Math.min(r, g, b);

    let h;
    let s;
    let l = (max + min) / 2;

    if (max === min) {
        h = s = 0; // achromatic
    }
    else {
        const d = max - min;
        s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
        switch (max) {
            case r: h = (g - b) / d + (g < b ? 6 : 0); break;
            case g: h = (b - r) / d + 2; break;
            case b: h = (r - g) / d + 4; break;
        }
        h /= 6;
    }
    return [h, s, l];
}

/**
 * Converts the given hex color string to the equivalent rgba color string.
 * @param  {string}       hex     A hex color string.
 * @param  {number = 100} opacity A percentage of the opacity.
 * @return {string}               A rgba color string.
 */
export function hexToRgba(hex: string, opacity: number = 100) {
    if ( !hex ) {
        return 'rgba(221, 221, 221, 1)';
    }
    hex = hex.replace('#', '');
    const r = parseInt(hex.substring(0, 2), 16);
    const g = parseInt(hex.substring(2, 4), 16);
    const b = parseInt(hex.substring(4, 6), 16);
    return `rgba(${r}, ${g}, ${b}, ${opacity / 100})`;
}

/**
 * Converts hex or rgba color to hsl color.
 *
 * @param  {string} colorString A hex or rgb(a) color string.
 * @return {Array}             An array representing hsl color, [h, s, l].
 */
export function convertToHSL(colorString: string) {
    const rgba = colorString.indexOf('#') >= 0 ? hexToRgba(colorString) : colorString;
    const [r, g, b] = rgba.substring(rgba.indexOf('(') + 1, rgba.indexOf(')')).split(',').map(n => Number(n));
    return toHSL([r, g, b]);
}

