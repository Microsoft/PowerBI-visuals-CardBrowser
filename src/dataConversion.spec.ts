/*
 * Copyright 2018 Uncharted Software Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import * as utils from './utils';
import * as moment from 'moment';
import {
    convertToDocumentData,
    countDocuments,
} from './dataConversion';
import populateData from './test_data/testDataUtils';

jest.mock('moment');
jest.mock('./utils');

window['powerbi'] = {
    visuals: {
        SelectionIdBuilder: function () {
            this.withCategory = function () { return this; };
            this.createSelectionId = function () { return this; };
        }
    }
};

const DEFAULT_SETTINGS = {
    presentation: {
        wrap: true,
        height: 250,
        summaryUrl: true,
        dateFormat: 'MMM D, YYYY',
    },
    loadMoreData: {
        enabled: false,
        limit: 500
    },
};

describe('dataConversion', () => {
    describe('.convertToDocumentData', () => {
        it('should convert powerbi dataview to document data', () => {
            const { dataViews } = populateData([]);
            const docData = convertToDocumentData(<any>dataViews[0], DEFAULT_SETTINGS, {}, <any>{});
            expect(docData).toMatchSnapshot();
        });
    });
    describe('.countDocuments', () => {
        it('should cont documents', () => {
            const { dataViews } = populateData([]);
            const count = countDocuments(<any>dataViews[0]);
            expect(count).toBe(4);
        });
    });
});
