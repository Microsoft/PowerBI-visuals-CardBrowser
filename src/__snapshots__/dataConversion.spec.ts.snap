// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dataConversion .convertToDocumentData should convert powerbi dataview to document data 1`] = `
Object {
  "documentList": Array [
    Object {
      "author": 5,
      "cluster": 2935,
      "content": undefined,
      "document": 4423,
      "id": 4423,
      "identity": undefined,
      "imageUrl": 2935,
      "index": 0,
      "partContent": "Now is the time",
      "partIndex": 0,
      "score": "2017-01-01 20:43:39",
      "selectionId": null,
      "size": 5,
      "summary": undefined,
      "title": "Comments",
    },
    Object {
      "author": 2,
      "cluster": 8883,
      "content": undefined,
      "document": 15223,
      "id": 15223,
      "identity": undefined,
      "imageUrl": 8883,
      "index": 3,
      "partContent": "Free Market for Education",
      "partIndex": 0,
      "score": "2017-01-02 01:57:41",
      "selectionId": null,
      "size": 2,
      "summary": undefined,
      "title": "Free Market",
    },
    Object {
      "author": 64,
      "cluster": 1272,
      "content": undefined,
      "document": 211289,
      "id": 211289,
      "identity": undefined,
      "imageUrl": 1272,
      "index": 4,
      "partContent": "Washington",
      "partIndex": 0,
      "score": "2017-01-02 05:45:33",
      "selectionId": null,
      "size": 64,
      "summary": undefined,
      "title": "Democrats",
    },
    Object {
      "author": 2,
      "cluster": 13478,
      "content": undefined,
      "document": 211696,
      "id": 211696,
      "identity": undefined,
      "imageUrl": 13478,
      "index": 10,
      "partContent": "Sheltering in place",
      "partIndex": 0,
      "score": "2017-01-01 23:49:00",
      "selectionId": null,
      "size": 2,
      "summary": undefined,
      "title": "Sheltering",
    },
  ],
  "documents": Object {
    "15223": Object {
      "author": 2,
      "cluster": 8883,
      "content": undefined,
      "document": 15223,
      "id": 15223,
      "identity": undefined,
      "imageUrl": 8883,
      "index": 3,
      "partContent": "Free Market for Education",
      "partIndex": 0,
      "score": "2017-01-02 01:57:41",
      "selectionId": null,
      "size": 2,
      "summary": undefined,
      "title": "Free Market",
    },
    "211289": Object {
      "author": 64,
      "cluster": 1272,
      "content": undefined,
      "document": 211289,
      "id": 211289,
      "identity": undefined,
      "imageUrl": 1272,
      "index": 4,
      "partContent": "Washington",
      "partIndex": 0,
      "score": "2017-01-02 05:45:33",
      "selectionId": null,
      "size": 64,
      "summary": undefined,
      "title": "Democrats",
    },
    "211696": Object {
      "author": 2,
      "cluster": 13478,
      "content": undefined,
      "document": 211696,
      "id": 211696,
      "identity": undefined,
      "imageUrl": 13478,
      "index": 10,
      "partContent": "Sheltering in place",
      "partIndex": 0,
      "score": "2017-01-01 23:49:00",
      "selectionId": null,
      "size": 2,
      "summary": undefined,
      "title": "Sheltering",
    },
    "4423": Object {
      "author": 5,
      "cluster": 2935,
      "content": undefined,
      "document": 4423,
      "id": 4423,
      "identity": undefined,
      "imageUrl": 2935,
      "index": 0,
      "partContent": "Now is the time",
      "partIndex": 0,
      "score": "2017-01-01 20:43:39",
      "selectionId": null,
      "size": 5,
      "summary": undefined,
      "title": "Comments",
    },
  },
}
`;
